src/primitive_debug.cc

```diff
@@ -60,9 +54,40 @@ PRIMITIVE(object_histogram) {
     } else if (is_string(object) && !String::cast(object)->content_on_heap()) {
       size += String::cast(object)->length() + 1;
     }
-    capture.data[class_index * UINT32_PER_ENTRY + 0] += 1;
-    capture.data[class_index * UINT32_PER_ENTRY + 1] += size;
+    data[class_index].count++;
+    data[class_index].size += size;
   });
+  int non_trivial_entries = 0;
+  uword bytes_needed = 0;
+  for (int i = 0; i < length; i++) {
+    if (data[i].size > 0) {
+      non_trivial_entries++;
+      bytes_needed += Encoder::bytes_to_encode(i);
+      bytes_needed += Encoder::bytes_to_encode(data[i].count);
+      bytes_needed += Encoder::bytes_to_encode(data[i].size);
+    }
+  }
+  MallocedBuffer encoding_buffer(100 + bytes_needed);
+  if (!encoding_buffer.has_content()) MALLOC_FAILED;
+
+  ProgramOrientedEncoder encoder(program, &encoding_buffer);
```

Comment:
Maybe it would be easier just do the encoding twice, the first time with a zero-length buffer. We count the number of bytes in the encoder even after overflow.